---
title: "qtip_paper"
output:
  html_document:
    theme: united
    toc: yes
  pdf_document:
    toc: yes
---

This R markdown notebook depends on the existence of certain summary files from the experiments in the [`qsim-experiments` repo](https://github.com/BenLangmead/qsim-experiments).  It is meant to be executed from the `experiments` subdirectory of the `qsim-experiments` working set.  In particular, the notebook expects that these files are already present:

* `simulated_reads/summary/overall.csv`
* `real_data/perf.csv`
* `real_data/overall.csv`
* `platinum/ERR194147.csv`

See the `README.md` in the root of the `qsim-experiments` repo for instructions on how to generate those files.

### Load R libraries:

```{r libraries}
# Might require installation using install.packages()
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(data.table)
library(grid)
library(gridExtra)
library(xtable)
library(knitr)
library(tables)
# setwd('~/git/qtip-experiments/experiments/')
```

### Load/create R object for table

```{r read_table}
m <- read.table('simulated_reads/summary/overall.csv', sep=',', quote='', comment.char='', header=T)
m$aligner <- factor(m$aligner)
m$species <- factor(m$species)
m$sensitivity <- factor(m$sensitivity)
m$sim <- factor(m$sim)
m$aligner2 <- factor(ifelse(m$local,
                            ifelse(m$aligner == 'bt2', 'bt2 local', as.character(m$aligner)),
                            ifelse(m$aligner == 'bt2', 'bt2 e2e',   as.character(m$aligner))))
```

### Log mod scaling

```{r m_formatting}
mfmt <- function(n) { paste(n/1e6, 'M', sep='') }
```

```{r log_scaling}
logmod <- function(x) {
  sign(x) * log10(abs(x) + 1)
}

inv_logmod <- function(x) {
  sx <- sign(x)
  return(sx * (10.0 ** (sx * x) - 1))
}

logmod_breaks <- c(-100000, -10000, -1000, -100, -10, -1, 1, 10, 100, 1000, 10000, 100000)

logmod_trans <- function() {
  trans_new(paste0("logmod"), logmod, inv_logmod)
}
```

### Multi-plotting

```{r grid_arrange_shared}
# Borrowed from: https://rpubs.com/sjackman/grid_arrange_shared_legend
# Thanks to Shaun Jackman
grid_arrange_shared_legend <- function(...) {
    plots <- list(...)
    g <- ggplotGrob(plots[[1]] + theme(legend.position="bottom"))$grobs
    legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
    lheight <- sum(legend$height)
    grid.arrange(
        do.call(arrangeGrob, lapply(plots, function(x)
            x + theme(legend.position="none"))),
        legend,
        ncol = 1,
        heights = unit.c(unit(1, "npc") - lheight, lheight))
}

grid_arrange_shared_legend2 <- function(..., ncol = length(list(...)), nrow = 1, position = c("bottom", "right")) {

  plots <- list(...)
  position <- match.arg(position)
  g <- ggplotGrob(plots[[1]] + theme(legend.position = position))$grobs
  legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
  lheight <- sum(legend$height)
  lwidth <- sum(legend$width)
  gl <- lapply(plots, function(x) x + theme(legend.position="none"))
  gl <- c(gl, ncol = ncol, nrow = nrow)

  combined <- switch(position,
                     "bottom" = arrangeGrob(do.call(arrangeGrob, gl),
                                            legend,
                                            ncol = 1,
                                            heights = unit.c(unit(1, "npc") - lheight, lheight)),
                     "right" = arrangeGrob(do.call(arrangeGrob, gl),
                                           legend,
                                           ncol = 2,
                                           widths = unit.c(unit(1, "npc") - lwidth, lwidth)))

  grid.newpage()
  grid.draw(combined)

  # return gtable invisibly
  invisible(combined)

}
```

### Expand ROCs into CID and CSE curves

```{r roc_handling}
roc_string_to_table <- function(roc_string) {
  conn <- textConnection(gsub(';', '\n', roc_string), local=T)
  roct <- read.table(conn, sep=':')
  colnames(roct) <- c('mapq', 'cor', 'incor')
  stopifnot(!is.unsorted(rev(roct$mapq)))
  roct$n = roct$cor + roct$incor
  roct$pcor = 1 - (10 ** (roct$mapq/-10))
  roct$emp_pcor = roct$cor / roct$n
  close(conn)
  roct
}

tables_to_cid <- function(rl, rr) {
  stopifnot(sum(rl$n) == sum(rr$n))
  stopifnot(sum(rl$incor) == sum(rr$incor))
  rlci <- rep(rl$incor/rl$n, rl$n) 
  rrci <- rep(rr$incor/rr$n, rr$n)
  stopifnot(sum(rlci) - sum(rrci) < (1e-6 * nrow(rl)))
  cumsum(rlci - rrci)
}

# relative change in area under CID
rca <- function(r_predict, r_orig) {
  tot <- sum(cumsum(rep(r_orig$incor/r_orig$n, r_orig$n)))
  (sum(cumsum(rep(r_predict$incor/r_predict$n, r_predict$n))) - tot) / tot
}

tables_to_cse <- function(rl, rr) {
  stopifnot(sum(rl$n) == sum(rr$n))
  stopifnot(sum(rl$incor) == sum(rr$incor))
  incor_se_l <- rl$pcor**2 * rl$incor
  cor_se_l <- (1 - rl$pcor)**2 * rl$cor
  incor_se_r <- rr$pcor**2 * rr$incor
  cor_se_r <- (1 - rr$pcor)**2 * rr$cor
  cumsum(rep((incor_se_l + cor_se_l)/rl$n, rl$n) -
         rep((incor_se_r + cor_se_r)/rr$n, rr$n))
}


# relative change in squared error
rce <- function(r_predict, r_orig) {
  # total error attributable to incorrects in each stratum
  incor_se_orig <- r_orig$pcor**2 * r_orig$incor
  incor_se_predict <- r_predict$pcor**2 * r_predict$incor
  # total error attributable to corrects in each stratum
  cor_se_orig <- (1 - r_orig$pcor)**2 * r_orig$cor
  cor_se_predict <- (1 - r_predict$pcor)**2 * r_predict$cor
  # total error
  err_orig <- sum(incor_se_orig + cor_se_orig)
  err_predict <- sum(incor_se_predict + cor_se_predict)
  (err_predict - err_orig) / err_orig
}

inflection_subsampler <- function(vec, length_out=1000) {
  pts <- union(1 + which(diff(vec, differences=2) > 1e-6),
               round(seq(1, length(vec), length.out=length_out)))
  data.frame(x=pts, y=vec[pts])
}

explode_curves <- function(tab, cid, length_out=1000) {
  dfs <- list()
  tables_to <- if(cid) tables_to_cid else tables_to_cse
  for(i in 1:nrow(tab)) {
    rl <- roc_string_to_table(tab$roc_round[i])
    rr <- roc_string_to_table(tab$roc_orig[i])
    vec <- inflection_subsampler(tables_to(rl, rr), length_out=length_out)
    dfs[[i]] <- suppressWarnings(cbind(x=vec$x, y=vec$y, tab[i,]))
  }
  rbindlist(dfs)
}
```

```{r diagnostic_plots}
trials_plot <- function(m_filt, want_cid, xlab) {
  mtmp <- explode_curves(m_filt, want_cid)
  mx <- max(abs(mtmp$y))
  ggplot(mtmp, aes(x=x, y=y, color=factor(trial_no), linetype=mapq_included)) +
    geom_abline(intercept = 0, slope = 0) +
    coord_trans(y = logmod_trans()) +
    scale_y_continuous(breaks=logmod_breaks, limits=c(-1.05*mx, 1.05*mx)) +
    labs(x=xlab, y=if(want_cid) "CID" else "CSED") +
    geom_line() +
    theme_bw() +
    theme(legend.title=element_blank(), axis.title.x = element_text(vjust=-0.4))
}

subsample_trials_plot <- function(m_filt, want_cid) {
  f_te <- trials_plot(m_filt %>% filter(!training), want_cid, "test")
  f_tr <- trials_plot(m_filt %>% filter( training), want_cid, "training")
  grid_arrange_shared_legend(f_te, f_tr)
}
```

### Simulated reads

#### Tables

Start with tables:

```{r various_lengths_table}
various_lengths_table <- function(m, want_training, ...) {
  mtmp <- m %>% filter(training == want_training & name == 'ill_various_length')
  mtmp <- subset(mtmp, select=c(paired,
                                readlen,
                                local,
                                overall_test_auc_diff_pct_round,
                                overall_test_mse_diff_pct_round))
  colnames(mtmp) <- c('paired', 'readlen', 'local', 'RCA', 'RCE')
  mtmp$paired <- factor(mtmp$paired, labels=c('Unpaired', 'Paired'))
  mtmp$local <- factor(mtmp$local, labels=c('End-to-end', 'Local'))
  mtmp$readlen <- factor(mtmp$readlen)
  tabular((Heading() * paired * Heading("Read length") * readlen) ~ (Format(sprintf("%.2f")) * Justify(r) * Heading() * local * (RCA + RCE) * (mean + sd)), data=mtmp)
}

fn <- 'various_lengths.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(various_lengths_table(m, F)), file=fn, append=T)
write("\\caption{Relative change in area under CID (RCA) and relative change in sum of squared error (RCE) when running Qtip and Bowtie 2 on Mason-simulated Illumina-like samples of various lengths.  Relative change is expressed as a percent.  Each sample consists of 4 million reads/pairs.  Samples are either unpaired and paired-end, and Bowtie 2 is run in either end-to-end or local alignment mode as indicated.  Results are means and standard deviations over \\numtrials random trials, repeated starting from the input modeling step.}", file=fn, append=T)
write("\\label{length_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

```{r various_sensitivities_table}
various_sensitivities_table <- function(m, want_training, ...) {
  mtmp <- m %>% filter(training == want_training & name == 'various_sensitivities')
  mtmp <- subset(mtmp, select=c(paired,
                                readlen,
                                local,
                                sensitivity,
                                overall_test_auc_diff_pct_round,
                                overall_test_mse_diff_pct_round))
  colnames(mtmp) <- c('paired', 'readlen', 'local', 'sensitivity', 'RCA', 'RCE')
  mtmp$paired <- factor(mtmp$paired, labels=c('Unpaired', 'Paired'))
  mtmp$local <- factor(mtmp$local, labels=c('End-to-end', 'Local'))
  mtmp$readlen <- factor(mtmp$readlen)
  sens <- factor(gsub('[l0-9].*', '', mtmp$sensitivity))
  sens <- factor(sens, labels=c('--fast', '--sensitive', '--very-fast', '--very-sensitive'))
  sens <- relevel(sens, '--sensitive')
  sens <- relevel(sens, '--fast')
  sens <- relevel(sens, '--very-fast')
  tabular((Heading() * paired * Heading('Length') * readlen * Heading("Sensitivity") * sens) ~ (Format(sprintf("%.2f")) * Justify(r) * Heading() * local * (RCA + RCE) * (mean + sd)), data=mtmp)
}

fn <- 'various_sensitivities.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(various_sensitivities_table(m, F)), file=fn, append=T)
write("\\caption{Relative change in area under CID (RCA) and relative change in sum of squared error (RCE) when running Qtip and Bowtie 2 at various levels of sensitivity on Mason-simulated Illumina-like samples.  Relative change is expressed as a percent.  Each simulated sample consists of 4 million reads or pairs.  Bowtie 2 is run in either end-to-end or local alignment mode as indicated.  Results are means and standard deviations over \\numtrials random trials, repeated starting from the input modeling step.}", file=fn, append=T)
write("\\label{sens_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

```{r various_genomes_table}
various_genomes_table <- function(m, want_training, want_series, ...) {
  mtmp <- m %>% filter(training == want_training & name == want_series)
  mtmp <- subset(mtmp, select=c(aligner,
                                species,
                                paired,
                                readlen,
                                overall_test_auc_diff_pct_round,
                                overall_test_mse_diff_pct_round))
  levels(mtmp$aligner)[levels(mtmp$aligner)=="bt2"] <- "Bowtie 2"
  levels(mtmp$aligner)[levels(mtmp$aligner)=="bwamem"] <- "BWA-MEM"
  levels(mtmp$aligner)[levels(mtmp$aligner)=="snap"] <- "SNAP"
  levels(mtmp$species)[levels(mtmp$species)=="hg19"] <- "GRCh37"
  levels(mtmp$species)[levels(mtmp$species)=="hg38"] <- "GRCh38"
  levels(mtmp$species)[levels(mtmp$species)=="mm"] <- "Mouse"  # GRCm38
  levels(mtmp$species)[levels(mtmp$species)=="zm"] <- "Zea mays"  # AGPv3
  colnames(mtmp) <- c('aligner', 'species', 'paired', 'readlen', 'RCA', 'RCE')
  tabular(((Heading("Unpaired") * (!paired) + Heading("Paired") * paired) * Heading() * species * Heading() * aligner) ~ (Format(sprintf("%.2f")) * (Heading("100 nt") * (readlen == 100) + Heading("250 nt") * (readlen == 250)) * (RCA + RCE) * (mean + sd)), data=mtmp)
}

fn <- 'various_genomes.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(various_genomes_table(m, F, 'various_genomes')), file=fn, append=T)
write("\\caption{Relative change in area under CID (RCA) and relative change in sum of squared error (RCE) for various aligners and reference genomes, expressed as percent change.  The experiments used 100 nt or 250 nt reads, and unpaired or paired-end reads, as indicated.  Results are means and standard deviations over \\numtrials random trials, repeated starting from the input modeling step.}", file=fn, append=T)
write("\\label{genomes_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

```{r various_genomes_chm1_table}
fn <- 'various_genomes_chm1.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(various_genomes_table(m, F, 'various_genomes_chm1')), file=fn, append=T)
write("\\caption{Relative change in area under CID (RCA) and relative change in sum of squared error (RCE) for various aligners and reference genomes, expressed as percent change.  The experiments simulated the input reads from a combination of the appropriate human reference genome (GRCh37 or GRCh38) and additional contigs containing sequence from CHM1 that does not align well to the reference. The experiments used 100 nt or 250 nt reads, and unpaired or paired-end reads, as indicated.  Results are means and standard deviations over \\numtrials random trials, repeated starting from the input modeling step.}", file=fn, append=T)
write("\\label{genomes_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

```{r various_simulators_table}
various_simulators_table <- function(m, want_training, ...) {
  mtmp <- m %>% filter(training == want_training & name == 'various_simulators')
  mtmp <- subset(mtmp, select=c(sim,
                                paired,
                                readlen,
                                overall_test_auc_diff_pct_round,
                                overall_test_mse_diff_pct_round))
  levels(mtmp$sim)[levels(mtmp$sim)=="art"] <- "Art"
  levels(mtmp$sim)[levels(mtmp$sim)=="mason"] <- "Mason"
  levels(mtmp$sim)[levels(mtmp$sim)=="wgsim"] <- "wgsim"
  colnames(mtmp) <- c('simulator', 'paired', 'readlen', 'RCA', 'RCE')
  tabular((Heading("Unpaired") * (!paired) + Heading("Paired") * paired) *
          (Heading("100 nt") * (readlen == 100) + Heading("250 nt") * (readlen == 250)) *
          Heading() * simulator ~ (Format(sprintf("%.2f")) * (RCA + RCE) * (mean + sd)), data=mtmp)
}

fn <- 'various_simulators.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(various_simulators_table(m, F)), file=fn, append=T)
write("\\caption{Relative change in area under CID (RCA) and relative change in sum of squared error (RCE) for samples simulated by various simulators.  Bowtie 2 end-to-end alignment was used in all cases. Relative change is expressed as a percent.  The experiments used 100 nt or 250 nt reads, and unpaired or paired-end reads, as indicated. Results are means and standard deviations over \\numtrials random trials, repeated starting from the input modeling step.}", file=fn, append=T)
write("\\label{simulators_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

#### Various read lengths

```{r various_lengths_plot}

various_lengths_plot <- function(
  want_paired,
  want_training,
  want_cid,
  want_local,
  want_trial=1)
{
  xlab <- paste0(if(want_paired) "Paired " else "Unpaired ",
                 if(want_local) "local" else "end-to-end")
  mtmp <- m %>% filter(name == 'ill_various_length' &
                       paired == want_paired &
                       local == want_local &
                       training == want_training &
                       trial_no == want_trial)
  mtmp <- explode_curves(mtmp, want_cid)
  mx <- max(abs(mtmp$y))
  ggplot(mtmp, aes(x=x, y=y, color=factor(readlen))) +
    geom_abline(intercept = 0, slope = 0) +
    coord_trans(y = logmod_trans()) +
    scale_y_continuous(breaks=logmod_breaks, limits=c(-1.05*mx, 1.05*mx)) +
    scale_x_continuous(labels=mfmt) +
    labs(x=xlab, y="") +
    geom_line() +
    theme_bw() +
    guides(colour = guide_legend(override.aes = list(size=1))) +
    theme(legend.title=element_blank(), legend.key = element_blank(),
          legend.text=element_text(size=12))
}
```

```{r various_lengths_plot_cid_test}

## Test ##

# params: paired, training, cid, local
p_vl_ueie <- various_lengths_plot(F, F, T, F)
p_vl_ueil <- various_lengths_plot(F, F, T, T)
p_vl_peie <- various_lengths_plot(T, F, T, F)
p_vl_peil <- various_lengths_plot(T, F, T, T)

# Write to notebook
grid_arrange_shared_legend(p_vl_ueie,
                           p_vl_ueil,
                           p_vl_peie,
                           p_vl_peil)

# Write to PDF
pdf(file="various_lengths_cid_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vl_ueie,
                           p_vl_ueil,
                           p_vl_peie,
                           p_vl_peil)
dev.off()
```

```{r various_lengths_plot_cse_test}

## Test ##

# params: paired, training, cid, local
p_vl_ueee <- various_lengths_plot(F, F, F, F)
p_vl_ueel <- various_lengths_plot(F, F, F, T)
p_vl_peee <- various_lengths_plot(T, F, F, F)
p_vl_peel <- various_lengths_plot(T, F, F, T)

# Write to notebook
grid_arrange_shared_legend(p_vl_ueee,
                           p_vl_ueel,
                           p_vl_peee,
                           p_vl_peel)

# Write to PDF
pdf(file="various_lengths_cse_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vl_ueee,
                           p_vl_ueel,
                           p_vl_peee,
                           p_vl_peel)
dev.off()
```

```{r various_lengths_plot_cid_train}

## Training ##

# params: paired, training, cid, local
p_vl_urie <- various_lengths_plot(F, T, T, F)
p_vl_uril <- various_lengths_plot(F, T, T, T)
p_vl_prie <- various_lengths_plot(T, T, T, F)
p_vl_pril <- various_lengths_plot(T, T, T, T)

# Write to notebook
grid_arrange_shared_legend(p_vl_urie,
                           p_vl_uril,
                           p_vl_prie,
                           p_vl_pril)

# Write to PDF
pdf(file="various_lengths_cid_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vl_urie,
                           p_vl_uril,
                           p_vl_prie,
                           p_vl_pril)
dev.off()
```

```{r various_lengths_plot_cse_train}

## Training ##

# params: paired, training, cid, local
p_vl_uree <- various_lengths_plot(F, T, F, F)
p_vl_urel <- various_lengths_plot(F, T, F, T)
p_vl_pree <- various_lengths_plot(T, T, F, F)
p_vl_prel <- various_lengths_plot(T, T, F, T)

# Write to notebook
grid_arrange_shared_legend(p_vl_uree,
                           p_vl_urel,
                           p_vl_pree,
                           p_vl_prel)

# Write to PDF
pdf(file="various_lengths_cse_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vl_uree,
                           p_vl_urel,
                           p_vl_pree,
                           p_vl_prel)
dev.off()
```

#### Genomes

```{r various_genomes_plot}
various_genomes_plot <- function(
  want_series,
  want_paired,
  want_readlen,
  want_training,
  want_cid,
  want_trial=1)
{
    xlab <- paste0(want_readlen, " bp ",
                   if(want_paired) "paired " else "unpaired ")
    mtmp <- m %>% filter(name == want_series &
                         paired == want_paired &
                         readlen == want_readlen &
                         training == want_training &
                         trial_no == want_trial)
    mtmp <- explode_curves(mtmp, want_cid)
    mx <- max(abs(mtmp$y))
    levels(mtmp$species)[levels(mtmp$species)=="hg38"] <- "Human (GRCh38)  "
    levels(mtmp$species)[levels(mtmp$species)=="hg19"] <- "Human (GRCh37)  "
    levels(mtmp$species)[levels(mtmp$species)=="mm"] <- "Mouse (GRCm38)  "
    levels(mtmp$species)[levels(mtmp$species)=="zm"] <- "Zea Mays (AGPv4)  "
    levels(mtmp$aligner)[levels(mtmp$aligner)=="bt2"] <- "Bowtie 2  "
    levels(mtmp$aligner)[levels(mtmp$aligner)=="bwamem"] <- "BWA-MEM  "
    levels(mtmp$aligner)[levels(mtmp$aligner)=="snap"] <- "SNAP  "
    ggplot(mtmp, aes(x=x, y=y, color=species, linetype=aligner)) +
      geom_abline(intercept = 0, slope = 0) +
      coord_trans(y = logmod_trans()) +
      scale_y_continuous(breaks=logmod_breaks, limits=c(-1.05*mx, 1.05*mx)) +
      scale_x_continuous(labels=mfmt) +
      labs(x=xlab, y="") +
      geom_line() +
      theme_bw() +
      guides(colour = guide_legend(override.aes = list(size=1))) +
      theme(legend.title=element_blank(), legend.key = element_blank(),
            legend.text=element_text(size=12))
}
```

```{r various_genomes_plot_cid_test}

## Test ##

# params: paired, readlen, training, cid
p_vg_u100ei <- various_genomes_plot('various_genomes', F, 100, F, T)
p_vg_u250ei <- various_genomes_plot('various_genomes', F, 250, F, T)
p_vg_p100ei <- various_genomes_plot('various_genomes', T, 100, F, T)
p_vg_p250ei <- various_genomes_plot('various_genomes', T, 250, F, T)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100ei,
                           p_vg_u250ei,
                           p_vg_p100ei,
                           p_vg_p250ei)

# Write to PDF
pdf(file="various_genomes_cid_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100ei,
                           p_vg_u250ei,
                           p_vg_p100ei,
                           p_vg_p250ei)
dev.off()
```

```{r various_genomes_chm1_plot_cid_test}

## Test ##

# params: paired, readlen, training, cid
p_vg_u100ei <- various_genomes_plot('various_genomes_chm1', F, 100, F, T)
p_vg_u250ei <- various_genomes_plot('various_genomes_chm1', F, 250, F, T)
p_vg_p100ei <- various_genomes_plot('various_genomes_chm1', T, 100, F, T)
p_vg_p250ei <- various_genomes_plot('various_genomes_chm1', T, 250, F, T)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100ei,
                           p_vg_u250ei,
                           p_vg_p100ei,
                           p_vg_p250ei)

# Write to PDF
pdf(file="various_genomes_chm1_cid_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100ei,
                           p_vg_u250ei,
                           p_vg_p100ei,
                           p_vg_p250ei)
dev.off()
```

```{r various_genomes_plot_cid_train}

## Training ##

# params: paired, readlen, training, cid
p_vg_u100ri <- various_genomes_plot('various_genomes', F, 100, T, T)
p_vg_u250ri <- various_genomes_plot('various_genomes', F, 250, T, T)
p_vg_p100ri <- various_genomes_plot('various_genomes', T, 100, T, T)
p_vg_p250ri <- various_genomes_plot('various_genomes', T, 250, T, T)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100ri,
                           p_vg_u250ri,
                           p_vg_p100ri,
                           p_vg_p250ri)

# Write to PDF
pdf(file="various_genomes_cid_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100ri,
                           p_vg_u250ri,
                           p_vg_p100ri,
                           p_vg_p250ri)
dev.off()
```

```{r various_genomes_chm1_plot_cid_train}

## Training ##

# params: paired, readlen, training, cid
p_vg_u100ri <- various_genomes_plot('various_genomes_chm1', F, 100, T, T)
p_vg_u250ri <- various_genomes_plot('various_genomes_chm1', F, 250, T, T)
p_vg_p100ri <- various_genomes_plot('various_genomes_chm1', T, 100, T, T)
p_vg_p250ri <- various_genomes_plot('various_genomes_chm1', T, 250, T, T)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100ri,
                           p_vg_u250ri,
                           p_vg_p100ri,
                           p_vg_p250ri)

# Write to PDF
pdf(file="various_genomes_chm1_cid_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100ri,
                           p_vg_u250ri,
                           p_vg_p100ri,
                           p_vg_p250ri)
dev.off()
```

```{r various_genomes_plot_cse_test}

## Test ##

# params: paired, readlen, training, cid
p_vg_u100ee <- various_genomes_plot('various_genomes', F, 100, F, F)
p_vg_u250ee <- various_genomes_plot('various_genomes', F, 250, F, F)
p_vg_p100ee <- various_genomes_plot('various_genomes', T, 100, F, F)
p_vg_p250ee <- various_genomes_plot('various_genomes', T, 250, F, F)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100ee,
                           p_vg_u250ee,
                           p_vg_p100ee,
                           p_vg_p250ee)

# Write to PDF
pdf(file="various_genomes_cse_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100ee,
                           p_vg_u250ee,
                           p_vg_p100ee,
                           p_vg_p250ee)
dev.off()
```

```{r various_genomes_chm1_plot_cse_test}

## Test ##

# params: paired, readlen, training, cid
p_vg_u100ee <- various_genomes_plot('various_genomes_chm1', F, 100, F, F)
p_vg_u250ee <- various_genomes_plot('various_genomes_chm1', F, 250, F, F)
p_vg_p100ee <- various_genomes_plot('various_genomes_chm1', T, 100, F, F)
p_vg_p250ee <- various_genomes_plot('various_genomes_chm1', T, 250, F, F)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100ee,
                           p_vg_u250ee,
                           p_vg_p100ee,
                           p_vg_p250ee)

# Write to PDF
pdf(file="various_genomes_chm1_cse_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100ee,
                           p_vg_u250ee,
                           p_vg_p100ee,
                           p_vg_p250ee)
dev.off()
```

```{r various_genomes_plot_cse_train}

## Training ##

# params: paired, readlen, training, cid
p_vg_u100re <- various_genomes_plot('various_genomes', F, 100, T, F)
p_vg_u250re <- various_genomes_plot('various_genomes', F, 250, T, F)
p_vg_p100re <- various_genomes_plot('various_genomes', T, 100, T, F)
p_vg_p250re <- various_genomes_plot('various_genomes', T, 250, T, F)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100re,
                           p_vg_u250re,
                           p_vg_p100re,
                           p_vg_p250re)

# Write to PDF
pdf(file="various_genomes_cse_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100re,
                           p_vg_u250re,
                           p_vg_p100re,
                           p_vg_p250re)
dev.off()
```

```{r various_genomes_chm1_plot_cse_train}

## Training ##

# params: paired, readlen, training, cid
p_vg_u100re <- various_genomes_plot('various_genomes_chm1', F, 100, T, F)
p_vg_u250re <- various_genomes_plot('various_genomes_chm1', F, 250, T, F)
p_vg_p100re <- various_genomes_plot('various_genomes_chm1', T, 100, T, F)
p_vg_p250re <- various_genomes_plot('various_genomes_chm1', T, 250, T, F)

# Write to notebook
grid_arrange_shared_legend(p_vg_u100re,
                           p_vg_u250re,
                           p_vg_p100re,
                           p_vg_p250re)

# Write to PDF
pdf(file="various_genomes_chm1_cse_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vg_u100re,
                           p_vg_u250re,
                           p_vg_p100re,
                           p_vg_p250re)
dev.off()
```

##### Diagnostics

Zooming in on BWA-MEM 100 bp human CID:

```{r various_genomes_plot_bwa_100_unp_hg}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'hg38' & !paired & readlen == 100 & aligner == 'bwamem'), T)
```

Now 250 bp:

```{r various_genomes_plot_bwa_250_unp_mm}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'hg38' & !paired & readlen == 250 & aligner == 'bwamem'), T)
```

Now corn 100 bp unpaired:

```{r various_genomes_plot_bwa_100_unp_zm}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'zm' & !paired & readlen == 100 & aligner == 'bwamem'), T)
```

And 250 bp unpaired:

```{r various_genomes_plot_bwa_250_unp_zm}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'zm' & !paired & readlen == 250 & aligner == 'bwamem'), T)
```

```{r r12_snap100_hg_mason_ill_hg_100_diag}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'hg38' & paired & readlen == 100 & aligner == 'snap'), T)
```

```{r r12_snap250_hg_mason_ill_hg_250_diag}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'hg38' & paired & readlen == 250 & aligner == 'snap'), T)
```

```{r r12_snap250_zm_mason_ill_zm_250_diag}
subsample_trials_plot(m %>% filter(name == 'various_genomes' & species == 'zm' & paired & readlen == 250 & aligner == 'snap'), T)
```

#### Various sensitivities:

```{r various_sensitivities_plot}
various_sensitivities_plot <- function(
  want_paired,
  want_local,
  want_readlen,
  want_training,
  want_cid,
  want_trial=1)
{
  xlab <- paste0(want_readlen, " bp ",
                 if(want_paired) "paired " else "unpaired ",
                 if(want_local) "local" else "end-to-end")
  mtmp <- m %>% filter(name == 'various_sensitivities' &
                      paired == want_paired &
                      local == want_local &
                      readlen == want_readlen &
                      training == want_training &
                      trial_no == want_trial)
  levels(mtmp$sensitivity)[levels(mtmp$sensitivity)=="vs"] <- "--very-sensitive  "
  levels(mtmp$sensitivity)[levels(mtmp$sensitivity)=="s"] <- "--sensitive  "
  levels(mtmp$sensitivity)[levels(mtmp$sensitivity)=="f"] <- "--fast  "
  levels(mtmp$sensitivity)[levels(mtmp$sensitivity)=="vf"] <- "--very-fast  "
  mtmp$sensitivity <- relevel(mtmp$sensitivity, "--sensitive  ")
  mtmp$sensitivity <- relevel(mtmp$sensitivity, "--fast  ")
  mtmp$sensitivity <- relevel(mtmp$sensitivity, "--very-fast  ")
  mtmp <- explode_curves(mtmp, want_cid)
  mx <- max(abs(mtmp$y))
  ggplot(mtmp, aes(x=x, y=y, color=sensitivity)) +
    geom_abline(intercept = 0, slope = 0) +
    coord_trans(y = logmod_trans()) +
    scale_y_continuous(breaks=logmod_breaks, limits=c(-1.05*mx, 1.05*mx)) +
    scale_x_continuous(labels=mfmt) +
    labs(x=xlab, y="") +
    geom_line() +
    theme_bw() +
    guides(colour = guide_legend(override.aes = list(size=1))) +
    theme(legend.title=element_blank(), legend.key = element_blank(),
          legend.text=element_text(size=12))
}
```

```{r various_sensitivities_plot_cid_test}

## Test ##

# params: paired, local, readlen, training, cid
p_vs_u100ei <- various_sensitivities_plot(F, F, 100, F, T)
p_vs_u250ei <- various_sensitivities_plot(F, T, 100, F, T)
p_vs_p100ei <- various_sensitivities_plot(T, F, 100, F, T)
p_vs_p250ei <- various_sensitivities_plot(T, T, 100, F, T)

# Write to notebook
grid_arrange_shared_legend(p_vs_u100ei,
                           p_vs_u250ei,
                           p_vs_p100ei,
                           p_vs_p250ei)

# Write to PDF
pdf(file="various_sensitivities_cid_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vs_u100ei,
                           p_vs_u250ei,
                           p_vs_p100ei,
                           p_vs_p250ei)
dev.off()
```

```{r various_sensitivities_plot_cid_train}

## Training ##

# params: paired, local, readlen, training, cid
p_vs_u100ri <- various_sensitivities_plot(F, F, 100, T, T)
p_vs_u250ri <- various_sensitivities_plot(F, T, 100, T, T)
p_vs_p100ri <- various_sensitivities_plot(T, F, 100, T, T)
p_vs_p250ri <- various_sensitivities_plot(T, T, 100, T, T)

# Write to notebook
grid_arrange_shared_legend(p_vs_u100ri,
                           p_vs_u250ri,
                           p_vs_p100ri,
                           p_vs_p250ri)

# Write to PDF
pdf(file="various_sensitivities_cid_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vs_u100ri,
                           p_vs_u250ri,
                           p_vs_p100ri,
                           p_vs_p250ri)
dev.off()
```

```{r various_sensitivities_plot_cse_test}

## Test ##

# params: paired, local, readlen, training, cid
p_vs_u100ee <- various_sensitivities_plot(F, F, 100, F, F)
p_vs_u250ee <- various_sensitivities_plot(F, T, 100, F, F)
p_vs_p100ee <- various_sensitivities_plot(T, F, 100, F, F)
p_vs_p250ee <- various_sensitivities_plot(T, T, 100, F, F)

# Write to notebook
grid_arrange_shared_legend(p_vs_u100ee,
                           p_vs_u250ee,
                           p_vs_p100ee,
                           p_vs_p250ee)

# Write to PDF
pdf(file="various_sensitivities_cse_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vs_u100ee,
                           p_vs_u250ee,
                           p_vs_p100ee,
                           p_vs_p250ee)
dev.off()
```

```{r various_sensitivities_plot_cse_train}

## Training ##

# params: paired, local, readlen, training, cid
p_vs_u100re <- various_sensitivities_plot(F, F, 100, T, F)
p_vs_u250re <- various_sensitivities_plot(F, T, 100, T, F)
p_vs_p100re <- various_sensitivities_plot(T, F, 100, T, F)
p_vs_p250re <- various_sensitivities_plot(T, T, 100, T, F)

# Write to notebook
grid_arrange_shared_legend(p_vs_u100re,
                           p_vs_u250re,
                           p_vs_p100re,
                           p_vs_p250re)

# Write to PDF
pdf(file="various_sensitivities_cse_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vs_u100re,
                           p_vs_u250re,
                           p_vs_p100re,
                           p_vs_p250re)
dev.off()
```

#### Zooming in on r0_bt2vf_mason_ill_250:

```{r r0_bt2vf_mason_ill_250_diag}
subsample_trials_plot(m %>% filter(name == 'various_sensitivities' & !paired & readlen == 250 & sensitivity == 'vf'), F)
```

```{r r12_bt2vs250_mason_ill_250_diag}
subsample_trials_plot(m %>% filter(name == 'various_sensitivities' & paired & readlen == 250 & sensitivity == 'vs250'), T)
```

```{r r0_bt2vsl_mason_ill_250_diag}
subsample_trials_plot(m %>% filter(name == 'various_sensitivities' & !paired & readlen == 250 & sensitivity == 'vsl'), F)
```

#### Simulators

```{r various_simulators_plot}
various_simulators_plot <- function(
  want_paired,
  want_readlen,
  want_training,
  want_cid,
  want_trial=1)
{
    xlab <- paste0(want_readlen, " bp ",
                   if(want_paired) "paired " else "unpaired ")
    mtmp <- m %>% filter(name == 'various_simulators' &
                         paired == want_paired &
                         readlen == want_readlen &
                         training == want_training &
                         trial_no == want_trial)
    stopifnot(nrow(mtmp) == 3)
    mtmp <- explode_curves(mtmp, want_cid)
    mx <- max(abs(mtmp$y))
    levels(mtmp$sim)[levels(mtmp$sim)=="art"] <- "Art  "
    levels(mtmp$sim)[levels(mtmp$sim)=="mason"] <- "Mason  "
    levels(mtmp$sim)[levels(mtmp$sim)=="wgsim"] <- "wgsim  "
    ggplot(mtmp, aes(x=x, y=y, color=sim)) +
      geom_abline(intercept = 0, slope = 0) +
      coord_trans(y = logmod_trans()) +
      scale_y_continuous(breaks=logmod_breaks, limits=c(-1.05*mx, 1.05*mx)) +
      scale_x_continuous(labels=mfmt) +
      labs(x=xlab, y="") +
      geom_line() +
      theme_bw() +
      guides(colour = guide_legend(override.aes = list(size=1))) +
      theme(legend.title=element_blank(), legend.key = element_blank(),
            legend.text=element_text(size=12))
}
```

```{r various_simulators_plot_cid_test}

## Test ##

# params: paired, local, readlen, training, cid
p_vi_u100ei <- various_simulators_plot(F, 100, F, T)
p_vi_u250ei <- various_simulators_plot(F, 250, F, T)
p_vi_p100ei <- various_simulators_plot(T, 100, F, T)
p_vi_p250ei <- various_simulators_plot(T, 250, F, T)

# Write to notebook
grid_arrange_shared_legend(p_vi_u100ei,
                           p_vi_u250ei,
                           p_vi_p100ei,
                           p_vi_p250ei)

# Write to PDF
pdf(file="various_simulators_cid_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vi_u100ei,
                           p_vi_u250ei,
                           p_vi_p100ei,
                           p_vi_p250ei)
dev.off()
```

```{r various_simulators_plot_cse_test}

## Test ##

# params: paired, local, readlen, training, cid
p_vi_u100ee <- various_simulators_plot(F, 100, F, F)
p_vi_u250ee <- various_simulators_plot(F, 250, F, F)
p_vi_p100ee <- various_simulators_plot(T, 100, F, F)
p_vi_p250ee <- various_simulators_plot(T, 250, F, F)

# Write to notebook
grid_arrange_shared_legend(p_vi_u100ee,
                           p_vi_u250ee,
                           p_vi_p100ee,
                           p_vi_p250ee)

# Write to PDF
pdf(file="various_simulators_cse_test.pdf", onefile=F)
grid_arrange_shared_legend(p_vi_u100ee,
                           p_vi_u250ee,
                           p_vi_p100ee,
                           p_vi_p250ee)
dev.off()
```

```{r various_simulators_plot_cid_train}

## Training ##

# params: paired, local, readlen, training, cid
p_vi_u100ri <- various_simulators_plot(F, 100, T, T)
p_vi_u250ri <- various_simulators_plot(F, 250, T, T)
p_vi_p100ri <- various_simulators_plot(T, 100, T, T)
p_vi_p250ri <- various_simulators_plot(T, 250, T, T)

# Write to notebook
grid_arrange_shared_legend(p_vi_u100ri,
                           p_vi_u250ri,
                           p_vi_p100ri,
                           p_vi_p250ri)

# Write to PDF
pdf(file="various_simulators_cid_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vi_u100ri,
                           p_vi_u250ri,
                           p_vi_p100ri,
                           p_vi_p250ri)
dev.off()
```

```{r various_simulators_plot_cse_train}

## Training ##

# params: paired, local, readlen, training, cid
p_vi_u100re <- various_simulators_plot(F, 100, T, F)
p_vi_u250re <- various_simulators_plot(F, 250, T, F)
p_vi_p100re <- various_simulators_plot(T, 100, T, F)
p_vi_p250re <- various_simulators_plot(T, 250, T, F)

# Write to notebook
grid_arrange_shared_legend(p_vi_u100re,
                           p_vi_u250re,
                           p_vi_p100re,
                           p_vi_p250re)

# Write to PDF
pdf(file="various_simulators_cse_training.pdf", onefile=F)
grid_arrange_shared_legend(p_vi_u100re,
                           p_vi_u250re,
                           p_vi_p100re,
                           p_vi_p250re)
dev.off()
```

##### Performance

```{r real_data_perf_tables}
# from qsim-experiments/experiments/real_data/perf_tabulate.py
perf <- read.table('real_data/perf_qtip162.csv', sep=',', quote='', comment.char='', header=T)
```

```{r}
real_data_perf_table <- function(dataset) {
  perftmp <- perf
  levels(perftmp$aligner)[levels(perftmp$aligner)=="bt2"] <- "Bowtie 2"
  levels(perftmp$aligner)[levels(perftmp$aligner)=="bwamem"] <- "BWA-MEM"
  levels(perftmp$aligner)[levels(perftmp$aligner)=="snap"] <- "SNAP"
  perftmp$ja <- perftmp$align_time / 60.0
  perftmp$oa <- perftmp$overall_time / 60.0
  perftmp$pc <- perftmp$peak_children / 1024.0 / 1024.0 / 1024.0
  perftmp$pw <- (perftmp$peak_wrapper + perftmp$peak_children) / 1024.0 / 1024.0 / 1024.0
  tabular(((Heading() * data) * (Heading("Unpaired") * (!paired) + Heading("Paired") * paired) *
            Heading() * aligner) ~
           (Heading() * identity * Justify(r) * Format(sprintf("%.2f")) *
              (Heading("Time") * ja +
               Heading("+Qtip") * oa +
               Heading("\\% inc") * pct_increase_a_to_o +
               Heading("Memory") * pc +
               Heading("+Qtip") * pw +
               Heading("\\% inc") * pct_increase_peak)),
          data=perftmp)
}
fn <- 'real_performance.tex_snippet'
write("\\begin{table}", file=fn)
write("% Add this line right after (or instead of) the first hline", file=fn, append=T)
write("%  &  &  & \\multicolumn{3}{c}{Time (minutes)} & \\multicolumn{3}{c}{Peak memory (gigabytes)} \\\\\\cline{4-6} \\cline{7-9}", file=fn, append=T)
write("\\centering", file=fn, append=T)
write("\\resizebox{\\columnwidth}{!}{%", file=fn, append=T)
capture.output(latex(real_data_perf_table()), file=fn, append=T)
write("}", file=fn, append=T)
write("\\caption{Overhead of the Qtip tool.  Measured as the increase in running time (left) and peak memory footprint (right) from when the aligner runs by itself (``Time'') to when the aligner runs in combination with Qtip (``+Qtip'').  ``\\% inc'' columns give the percent increase.  Times are in minutes and memory footprints are in gigabytes.}", file=fn, append=T)
write("\\label{real_performance_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

### Alignment error simulation

```{r}
ae <- read.table('alignment_err/results.csv', sep=',', quote='', comment.char='', header=T)
```

```{r}
ae$paired <- factor(ae$paired, labels=c('Unpaired', 'Paired'))
ae$species <- factor(ae$species, labels=c('GRCh38', 'GRCh38+CHM1', 'Mouse'))
ae$rdlen <- factor(ae$rdlen)
ae <- ae[ae$type != 1,]
ae$type <- factor(ae$type, labels=c('Category 1a', 'Category 1b', 'Category 2', 'Category 3'))
aetab <- tabular(
  (Heading() * species * Heading() * rdlen * Heading() * paired) ~
  (
    (Heading() * identity * Justify(r) * Format(sprintf("%.3f"))) *
    (Heading() * type * (Heading("Reads") * pct_total + Heading("Errors") * pct_error))
  ), data=ae)

fn <- 'aln_error.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(aetab), file=fn, append=T)
write("\\caption{Prevalence of errors of different categories when analyzing Mason-simulated samples using Bowtie 2. Columns labeled Reads give the percent of all the reads, both correctly and incorrectly aligned, having that error category. Columns labeled Errors give the percent of the errors having that error category.}", file=fn, append=T)
write("\\label{aln_error_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```

### Variant calling experiments

```{r}
mv <- read.table('platinum/ERR194147.csv', header=T, sep=',')

chr_table_sm <- function(want_chr) {
  df <- chr_table(want_chr)
  df[df$betas %in% c('4.000', '3.000', '2.000', '1.500', '1.000', '0.750', '0.500', '0.333', '0.250'),]
}

chr_table <- function(want_chr) {
  regime_cols <- colnames(mv)[grepl('^regime_[0-9]_[0-9][0-9][0-9]$', colnames(mv))]
  betas <- as.vector(sub('_', '.', sub('regime_', '', regime_cols)))
  pool_rows_nr <- mv$mapq_cutoff == '*' & mv$filt == 'cr_filt'
  m1 <- mv[pool_rows_nr & mv$regime != '*' & mv$chrom == want_chr,]
  f_cols <- colnames(mv)[grepl('^f_[0-9]_[0-9][0-9][0-9]$', colnames(mv))]
  tp_cols <- colnames(mv)[grepl('^tp_[0-9]_[0-9][0-9][0-9]$', colnames(mv))]
  fp_cols <- colnames(mv)[grepl('^fp_[0-9]_[0-9][0-9][0-9]$', colnames(mv))]
  f_input <- as.vector(t(m1[m1$regime == 'input', f_cols]))
  f_final <- as.vector(t(m1[m1$regime == 'final', f_cols]))
  tp_input <- as.vector(t(m1[m1$regime == 'input', tp_cols]))
  tp_final <- as.vector(t(m1[m1$regime == 'final', tp_cols]))
  fp_input <- as.vector(t(m1[m1$regime == 'input', fp_cols]))
  fp_final <- as.vector(t(m1[m1$regime == 'final', fp_cols]))
  qual_thresh_cols <- colnames(mv)[grepl('^qual_cutoff_[0-9]_[0-9][0-9][0-9]$', colnames(mv))]
  qual_thresh_input <- sprintf('%0.3g', as.vector(t(m1[m1$regime == 'input', qual_thresh_cols])))
  qual_thresh_final <- sprintf('%0.3g', as.vector(t(m1[m1$regime == 'final', qual_thresh_cols])))
  mapq_thresh_cols <- colnames(mv)[grepl('^mapq_cutoff_[0-9]_[0-9][0-9][0-9]$', colnames(mv))]
  mapq_thresh_input <- sprintf('%d', as.integer(as.vector(t(m1[m1$regime == 'input', mapq_thresh_cols]))))
  mapq_thresh_final <- sprintf('%d', as.integer(t(m1[m1$regime == 'final', mapq_thresh_cols])))
  change <- sprintf('%+0.1e', f_final - f_input)
  tp_change <- formatC(tp_final - tp_input, format='d', flag='+', big.mark=',')
  fp_change <- formatC(fp_final - fp_input, format='d', flag='+', big.mark=',')
  data.frame(betas=betas,
             f_input=sprintf('%0.4f', f_input),
             qt_input=qual_thresh_input,
             mt_input=mapq_thresh_input,
             buf=rep('', length(f_input)),
             f_final=sprintf('%0.4f', f_final),
             qt_final=qual_thresh_final,
             mt_final=mapq_thresh_final,
             buf=rep('', length(f_input)),
             f_change=change,
             tp_change=tp_change,
             fp_change=fp_change)
}
```

```{r}
print(xtable(chr_table_sm('W')), include.rownames=F)
```

<!---
\begin{table}[ht]
\centering
\resizebox{\columnwidth}{!}{%
\begin{tabular}{rrrrlrrrrrrrr}
  \hline
 & \multicolumn{3}{c}{Original} & & \multicolumn{3}{c}{Qtip} & & \multicolumn{3}{c}{$\Delta$ (Qtip $-$ Orig)} \\\cline{2-4}\cline{6-8}\cline{10-12}
\multicolumn{1}{c}{$\beta$} & \multicolumn{1}{c}{$F_\beta$} & QUAL & Q & & \multicolumn{1}{c}{$F_\beta$} & QUAL & Q & & \multicolumn{1}{c}{$F_\beta$} & \multicolumn{1}{c}{TP} & \multicolumn{1}{c}{FP} \\
  \hline
-->

<!---
   \hline
\end{tabular}
}
\caption{Single-nucleotide variant (SNV) $F_\beta$ scores for various $\beta$s with original mapping qualities and with Qtip-generated qualities. Paired-end reads from ERR194147, a female, were aligned with Bowtie 2 together with Qtip.  SNV variants were called with Freebayes for chromosomes 1-22 and X. Variant-quality (QUAL) and mapping-quality (Q) cutoffs yielding greatest $F_\beta$ score are reported. Platinum variants were used as the true callset.  Before calculating $F_\beta$, calls outside Platinum Genomes high-confidence regions were excluded.  Rightmost columns show differences in $F_\beta$, number of true positive SNVs, and number of false positive SNVs.}
\label{varcall_tab}
\end{table}
-->

### Target number of tandem reads

```{r tandem_target}
mtsu <- read.table('simulated_reads/big_experiments/r0_train_series.csv', sep=',', header=F)
mtsp <- read.table('simulated_reads/big_experiments/r12_train_series.csv', sep=',', header=F)
mtsu$paired <- F
mtsp$paired <- T

mts <- rbind(mtsu, mtsp)
colnames(mts) <- c('input', 'func', 'trial', 'rca', 'rcar', 'rce', 'rcer', 'paired')

tandem_target_plot <- function(m, want_paired) {
  tt <- paste(if(want_paired) 'Paired' else 'Unpaired', '100bp')
  mg <- m %>%
    filter(paired == want_paired) %>%
    group_by(input, func, paired) %>% 
    summarise(rce_med=median(rcer), rce_max=max(rcer), rce_min=min(rcer),
              rca_med=median(rcar), rca_max=max(rcar), rca_min=min(rcar))
  mg$reord <- 0
  mg$reord[mg$func ==  '15s'] <- 1
  mg$reord[mg$func ==  '30s'] <- 2
  mg$reord[mg$func ==  '45s'] <- 3
  mg$reord[mg$func ==   '1l'] <- 4
  mg$reord[mg$func ==   '3l'] <- 5
  mg$reord[mg$func ==   '5l'] <- 6
  mg$reord[mg$func == '050c'] <- 7
  mg$reord[mg$func == '100c'] <- 8
  mg$func <- reorder(mg$func, mg$reord)
  mg$func <- plyr::revalue(mg$func, c(
    '15s'='15 * sqrt(x)',
    '30s'='30 * sqrt(x)',
    '45s'='45 * sqrt(x)',
    '1l'='0.01 * x',
    '3l'='0.03 * x',
    '5l'='0.05 * x',
    '050c'='50,000',
    '100c'='100,000'))
  ggplot(mg, aes(x=input, color=func)) +
    geom_line(aes(y=rce_med)) +
    geom_point(aes(y=rce_med)) +
    theme_bw() +
    ggtitle(tt) +
    scale_x_continuous(labels = comma) +
    scale_colour_manual(breaks=mg$func, values = c('cadetblue4',
                                                   'chartreuse4',
                                                   'aquamarine3',
                                                   'red2',
                                                   'orangered1',
                                                   'darkred',
                                                   'dodgerblue3',
                                                   'blue')) +
    labs(x="# input reads", y="Median RCE, 10 trials") +
    theme(legend.title=element_blank(), legend.key = element_blank(),
          axis.text.x = element_text(angle = 45, hjust = 1))
}

# Write to PDF
tn_u <- tandem_target_plot(mts, F)
tn_p <- tandem_target_plot(mts, T)
pdf(file="tandem_num.pdf", onefile=F)
grid_arrange_shared_legend2(tn_u, tn_p, nrow=2, ncol=1)
dev.off()
```

### Feature importances

```{r feat_import_intro}
# Do various_genomes data, but caller specifies aligner and species
# We do 4 panels with 
feature_importances_plot_various_genomes <- function(
    want_aligner,
    want_model,
    want_readlen,
    top=6)
{
    aligner <- if(want_aligner == 'bt2') {
      'Bowtie 2'
    } else {
      if(want_aligner == 'bwamem') {
        'BWA-MEM'
      } else {
        'SNAP'
      }
    }
    type <- if     (want_model == 'c') { 'Conc' }
            else if(want_model == 'd') { 'Disc' }
            else if(want_model == 'b') { 'Bad-end' }
            else                       { 'Unp' }
    tt <- paste0(
      aligner, " ",
      type, ' ',
      #if(want_paired) "paired " else "unpaired ",
      want_readlen, 'nt'
      )
    mtmp <- m %>%
      dplyr::filter(name == 'various_genomes' &
             !training &
             aligner == want_aligner &
             readlen == want_readlen) %>%
      dplyr::select(trial_no, aligner, species, readlen, sim, feat) %>%
      dplyr::mutate(feat = as.character(feat)) %>%
      dplyr::group_by(trial_no, aligner, species, readlen, sim) %>%
      dplyr::mutate(feat = strsplit(feat, ';')) %>%
      tidyr::unnest() %>%
      tidyr::separate(feat, into=c('type', 'feature', 'importance'), sep=':') %>%
      dplyr::filter(type == want_model) %>%
      dplyr::mutate(feature = sub('_', '', feature), importance = as.numeric(importance))
    mtmp_top <- mtmp %>%
      group_by(aligner, readlen, sim, type, feature) %>%
      summarise(mn=mean(importance)) %>%
      top_n(n=top, wt=mn)
    mtmp_join = plyr::join(mtmp_top, mtmp, by=c('aligner', 'sim', 'readlen', 'type', 'feature'), type='inner')
    mtmp_join$feature <- reorder(mtmp_join$feature, -mtmp_join$mn)
  levels(mtmp_join$species)[levels(mtmp_join$species)=="hg19"] <- "GRCh37"
  levels(mtmp_join$species)[levels(mtmp_join$species)=="hg38"] <- "GRCh38"
  levels(mtmp_join$species)[levels(mtmp_join$species)=="mm"] <- "Mouse"  # GRCm38
  levels(mtmp_join$species)[levels(mtmp_join$species)=="zm"] <- "Zea mays"  # AGPv3
    if(want_aligner == 'bt2') {
        # Bowtie 2 paired
        mtmp_join$feature <- plyr::revalue(mtmp_join$feature, c(
          'fraglen'='Frag length',
          'ztz0'='Best score',
          'ztz1'='Score diff',
          'ztz2'='Edit diff',
          'ztz4'='Score diff other',
          'ztz6'='Conc score diff',
          'ztz7'='Conc edit diff',
          'ztz9'='% unique',
          'ztz10'='% repetitive',
          'ztz11'='Avg seed hits',
          'ztz13'='% unique stranded',
          'ztz15'='Avg seed hits stranded',
          'oztz2'='Edit diff other',
          'oztz13'='% unique stranded other'))
    } else if(want_aligner == 'bwamem') {
        mtmp_join$feature <- plyr::revalue(mtmp_join$feature, c(
          'fraglen'='Frag length',
          'ztz0'='Best score',
          'ztz1'='Best edit dist',
          'ztz2'='Score diff',
          'ztz3'='Score diff o',
          'oztz0'='Best score other',
          'oztz1'='Best edit dist other',
          'oztz2'='Score diff other',
          'oztz3'='Score diff o other',
          'ztz4'='Conc best score',
          'ztz5'='Conc score diff',
          'ztz6'='Seed length',
          'ztz7'='# tied for 2nd',
          'ztz8'='# tied for 2nd o',
          'ztz9'='# conc tied for 2nd',
          'oztz7'='# tied for 2nd other',
          'oztz8'='# tied for 2nd o other',
          'ztz10'='% repetitive',
          'ztz11'='% repetitive pair',
          'ztz12'='Seed coverage',
          'ztz13'='Seed coverage pair'))
    } else if(want_aligner == 'snap') {
        mtmp_join$feature <- plyr::revalue(mtmp_join$feature, c(
          'fraglen'='Frag length',
          'alqual'='Avg al base qual',
          'ztz0'='Best score',
          'ztz1'='Score diff',
          'ztz2'='Conc score diff',
          'ztz3'='Prob best',
          'oztz0'='Best score other',
          'oztz1'='Score diff other',
          'oztz2'='Conc score diff other',
          'oztz3'='Prob best other',
          'ztz4'='Pop seeds skip',
          'ztz5'='Pop seeds skip pair',
          'ztz6'='Min hits',
          'ztz7'='Min hits strand',
          'oztz6'='Min hits other',
          'oztz7'='Min hits strand other',
          'ztz8'='Hash misses',
          'oztz8'='Hash misses other',
          'ztz9'='Hash misses strand',
          'ztz10'='Hits per lookup',
          'ztz11'='Hits per lookup pair'))
    }
    ggplot(mtmp_join,
           aes(fill=species, x=feature, y=importance)) +
      geom_boxplot(outlier.size=NA) +
      labs(
           #x=paste('Top', top, 'features'),
           x='',
           y="Importance", title=tt) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1),
            legend.title=element_blank(),
            legend.key = element_blank())
}
fi_u_bt2_100 <- feature_importances_plot_various_genomes('bt2', 'u', 100)
fi_c_bt2_100 <- feature_importances_plot_various_genomes('bt2', 'c', 100)
fi_d_bt2_100 <- feature_importances_plot_various_genomes('bt2', 'd', 100)
fi_b_bt2_100 <- feature_importances_plot_various_genomes('bt2', 'b', 100)

pdf(file="feat_import_bt2_100.pdf", onefile=F)
grid_arrange_shared_legend(fi_u_bt2_100, fi_c_bt2_100,
                           fi_d_bt2_100, fi_b_bt2_100)
dev.off()

fi_u_bt2_250 <- feature_importances_plot_various_genomes('bt2', 'u', 250)
fi_c_bt2_250 <- feature_importances_plot_various_genomes('bt2', 'c', 250)
fi_d_bt2_250 <- feature_importances_plot_various_genomes('bt2', 'd', 250)
fi_b_bt2_250 <- feature_importances_plot_various_genomes('bt2', 'b', 250)

pdf(file="feat_import_bt2_250.pdf", onefile=F)
grid_arrange_shared_legend(fi_u_bt2_250, fi_c_bt2_250,
                           fi_d_bt2_250, fi_b_bt2_250)
dev.off()

fi_u_bwa_100 <- feature_importances_plot_various_genomes('bwamem', 'u', 100)
fi_c_bwa_100 <- feature_importances_plot_various_genomes('bwamem', 'c', 100)
fi_d_bwa_100 <- feature_importances_plot_various_genomes('bwamem', 'd', 100)

pdf(file="feat_import_bwa_100.pdf", onefile=F)
grid_arrange_shared_legend2(fi_u_bwa_100, fi_c_bwa_100, fi_d_bwa_100, nrow=2, ncol=2)
dev.off()

fi_u_bwa_250 <- feature_importances_plot_various_genomes('bwamem', 'u', 250)
fi_c_bwa_250 <- feature_importances_plot_various_genomes('bwamem', 'c', 250)
fi_d_bwa_250 <- feature_importances_plot_various_genomes('bwamem', 'd', 250)

pdf(file="feat_import_bwa_250.pdf", onefile=F)
grid_arrange_shared_legend2(fi_u_bwa_250, fi_c_bwa_250, fi_d_bwa_250, nrow=2, ncol=2)
dev.off()

fi_u_snap_100 <- feature_importances_plot_various_genomes('snap', 'u', 100)
fi_c_snap_100 <- feature_importances_plot_various_genomes('snap', 'c', 100)
fi_d_snap_100 <- feature_importances_plot_various_genomes('snap', 'd', 100)
fi_b_snap_100 <- feature_importances_plot_various_genomes('snap', 'b', 100)

pdf(file="feat_import_snap_100.pdf", onefile=F)
grid_arrange_shared_legend(fi_u_snap_100, fi_c_snap_100, fi_d_snap_100, fi_b_snap_100)
dev.off()

fi_u_snap_250 <- feature_importances_plot_various_genomes('snap', 'u', 250)
fi_c_snap_250 <- feature_importances_plot_various_genomes('snap', 'c', 250)
fi_d_snap_250 <- feature_importances_plot_various_genomes('snap', 'd', 250)
fi_b_snap_250 <- feature_importances_plot_various_genomes('snap', 'b', 250)

pdf(file="feat_import_snap_250.pdf", onefile=F)
grid_arrange_shared_legend(fi_u_snap_250, fi_c_snap_250, fi_d_snap_250, fi_b_snap_250)
dev.off()
```

### CHM1 comparisons

```{r various_genomes_chm1_comp_plot}
various_genomes_cmp_plot <- function(
  want_paired,
  want_readlen,
  want_training,
  want_cid,
  want_trial=1)
{
    xlab <- paste0(want_readlen, " bp ",
                   if(want_paired) "paired " else "unpaired ")
    mtmp <- m %>% filter(name == want_series &
                         paired == want_paired &
                         readlen == want_readlen &
                         training == want_training &
                         trial_no == want_trial)
    mtmp <- explode_curves(mtmp, want_cid)
    mx <- max(abs(mtmp$y))
    levels(mtmp$species)[levels(mtmp$species)=="hg38"] <- "Human (GRCh38)  "
    levels(mtmp$species)[levels(mtmp$species)=="hg19"] <- "Human (GRCh37)  "
    levels(mtmp$species)[levels(mtmp$species)=="mm"] <- "Mouse (GRCm38)  "
    levels(mtmp$species)[levels(mtmp$species)=="zm"] <- "Zea Mays (AGPv4)  "
    levels(mtmp$aligner)[levels(mtmp$aligner)=="bt2"] <- "Bowtie 2  "
    levels(mtmp$aligner)[levels(mtmp$aligner)=="bwamem"] <- "BWA-MEM  "
    levels(mtmp$aligner)[levels(mtmp$aligner)=="snap"] <- "SNAP  "
    ggplot(mtmp, aes(x=x, y=y, color=species, linetype=aligner)) +
      geom_abline(intercept = 0, slope = 0) +
      coord_trans(y = logmod_trans()) +
      scale_y_continuous(breaks=logmod_breaks, limits=c(-1.05*mx, 1.05*mx)) +
      scale_x_continuous(labels=mfmt) +
      labs(x=xlab, y="") +
      geom_line() +
      theme_bw() +
      guides(colour = guide_legend(override.aes = list(size=1))) +
      theme(legend.title=element_blank(), legend.key = element_blank(),
            legend.text=element_text(size=12))
}
```

```{r various_genomes_table}
various_genomes_cmp_table <- function(m, want_training, ...) {
  mtmp <- m %>% filter(training == want_training &
                         (name == 'various_genomes' | name == 'various_genomes_chm1'))
  mtmp <- mtmp %>% filter(species == 'hg19' | species == 'hg38')
  names <- mtmp$name
  mtmp <- subset(mtmp, select=c(aligner,
                                species,
                                paired,
                                readlen,
                                overall_test_auc_diff_pct_round,
                                overall_test_mse_diff_pct_round))
  mtmp$species <- factor(paste0(as.character(mtmp$species), ifelse(names == 'various_genomes_chm1', 'chm1', '')))
  levels(mtmp$aligner)[levels(mtmp$aligner)=="bt2"] <- "Bowtie 2"
  levels(mtmp$aligner)[levels(mtmp$aligner)=="bwamem"] <- "BWA-MEM"
  levels(mtmp$aligner)[levels(mtmp$aligner)=="snap"] <- "SNAP"
  levels(mtmp$species)[levels(mtmp$species)=="hg19"] <- "GRCh37"
  levels(mtmp$species)[levels(mtmp$species)=="hg38"] <- "GRCh38"
  levels(mtmp$species)[levels(mtmp$species)=="hg19chm1"] <- "GRCh37+CHM1"
  levels(mtmp$species)[levels(mtmp$species)=="hg38chm1"] <- "GRCh38+CHM1"
  colnames(mtmp) <- c('aligner', 'species', 'paired', 'readlen', 'RCA', 'RCE')
  tabular(((Heading("Unpaired") * (!paired) + Heading("Paired") * paired) * Heading() * aligner * Heading() * species) ~ (Format(sprintf("%.2f")) * (Heading("100 nt") * (readlen == 100) + Heading("250 nt") * (readlen == 250)) * (RCA + RCE) * (mean + sd)), data=mtmp)
}

fn <- 'various_genomes_cmp.tex_snippet'
write("\\begin{table}", file=fn)
write("\\centering", file=fn, append=T)
capture.output(latex(various_genomes_cmp_table(m, F)), file=fn, append=T)
write("\\caption{Relative change in area under CID (RCA) and relative change in sum of squared error (RCE) for various aligners and two versions of the human reference genome, expressed as percent change.  Some experiments simulated reads from just the reference (``GRCh37,'' ``GRCh38'') while other simulated from insertions in the CHM1 assembly relative to the reference assembly (``GRCh37+CHM1,'' ``GRCh38+CHM1'').  The experiments used 100 nt or 250 nt reads, and unpaired or paired-end reads, as indicated.  Results are means and standard deviations over \\numtrials random trials, repeated starting from the input modeling step.}", file=fn, append=T)
write("\\label{genomes_tab}", file=fn, append=T)
write("\\end{table}", file=fn, append=T)
```
